### ISSUE TITLE
glmmTMB cannot find betabinomal function

### ISSUE THREAD

**Issue by erinrickman** [glmmTMB/glmmTMB#988]
Hello, I am using a MacBook air with the M1 chip. 

I updated all my systems because I had logged on and updated the glmmTMB package, and my codes stopped running. I tried to install the older version but was hit with the issue of most Mac users with the M1 - I couldn't use dev tools to download packages or even install from source at all. Over a few hours, I fixed the issue by creating a makevars file. I was able to download the old package of glmmTMB. However, I can no longer use a betabinomal. 

Error in betabinomial(link = "logit") : 
  could not find function "betabinomial" 

The packages
install_version("glmmTMB", version = "1.1.6", repos = "http://cran.us.r-project.org")
install_version("TMB", version = "1.9.6", repos = "http://cran.us.r-project.org")
install_version("RcppEigen", version = "0.3.3.9.3", repos = "http://cran.us.r-project.org")
Here are my other loaded packages (I do a lot of exploration work)
require(tidyverse)
require(glmmTMB)
require(magrittr)
require(janitor)
require(rio)
require(DHARMa)
require(car)
require(effects)
require(emmeans)
require(sjPlot)
require(ggsignif)
require(ggtext)
require(dplyr)
require(ordinal)
require(hrbrthemes)
require(glmmTMB)
require(tidyverse)
require(magrittr)
require(data.table)
require(lme4)

I have tried all different package versions of glmmTMB, from the new ones to those 5 years old, but I cannot get this to function with a betabinomal. Does anyone know how to fix the issue? Here is my code, but this was working before today and I had not changed it:

proximityface <- glmmTMB::glmmTMB(cbind(convertedtime, convertedtotal) ~ condition*object_relation + (1|subject), 
                                 family = betabinomial(link = "logit"),
                          data = rawface[rawface$behavior == "proximity",])

My data has to have subjects counted, and due to animal illness, the data must be bound to make it weigh the same. 

If I try binomal(though not correct), I get this error message

Error in fitTMB(TMBStruc) : 
  negative log-likelihood is NaN at starting parameter values

The package will run other models, such as gaussian and Poisson 

Thanks!

**Comment by bbolker** [comment]
It's really hard to debug remotely.  

* When your codes "stopped running" with the new version of glmmTMB, what specifically happened? What version of the package were you running before? Did you switch to 1.1.8? It would probably be best in the long run to figure out how to run your code with the newer version of glmmTMB ... (I would have suspected the CppAD to TMBad switch in version 1.1.5, but you've reverted to a more recent version than that in your example above ...)
* `betabinomial` has been in glmmTMB for a long time, at least 5 years - I can't think of a reason why you wouldn't be able to find it, unless you had  forgotten to load the package (using `glmmTMB::glmmTMB(...)`, you could run the model without loading the package. Your code does say `require(glmmTMB)` above, though ...)
* when debugging it would be best if you can start from a *minimal* example, with a clean environment: for example, what do you get if you run the following code in a completely clean R session (no other packages loaded)?

```r
library(glmmTMB)
set.seed(101); dd <- data.frame(x = 1:10)
glmmTMB(cbind(x, 10-x) ~ 1, family = betabinomial, data = dd)
sessionInfo()
```

**Comment by erinrickman** [comment]
So, the issue I was originally getting when trying to run the model was this:

Error in betabinomial(link = "logit") : 
  could not find function "betabinomial"
In addition: Warning message:
In checkDepPackageVersion(dep_pkg = "TMB") :
  Package version inconsistency detected.
glmmTMB was built with TMB version 1.9.6
Current TMB version is 1.9.10
Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)

So, when I tried to install from source I was getting this message
It produced 193 warnings, which, i couldn't see them all, but the ones I could see as below, turned out to be an issue with the Mac M1 devices

ERROR: loading failed
* removing ‘/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/TMB’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/TMB’
Warning message:
In i.p(...) :
  installation of package ‘/var/folders/9c/tlz3nt3s2pvgzbmb55h9bq800000gn/T//RtmpbywRH5/remotes683427abc225/TMB’ had non-zero exit status

So, i created a Makevars file using this code:

# Newly installed Homebrew is located in
#  - /opt/homebrew for ARM Macs (M1 and its successors)
#  - /usr/local for Intel Macs
HOMEBREW_LOC=/opt/homebrew
# If you downloaded llvm manually above, replace with your chosen NEW_PATH/clang
LLVM_LOC=$(HOMEBREW_LOC)/opt/llvm
CC=$(LLVM_LOC)/bin/clang -fopenmp
CXX=$(LLVM_LOC)/bin/clang++ -fopenmp
# -O3 should be faster than -O2 (default) level optimisation ..
CFLAGS=-g -O3 -Wall -pedantic -std=gnu99 -mtune=native -pipe
CXXFLAGS=-g -O3 -Wall -pedantic -std=c++11 -mtune=native -pipe
LDFLAGS=-L$(HOMEBREW_LOC)/opt/gettext/lib -L$(LLVM_LOC)/lib -Wl,-rpath,$(LLVM_LOC)/lib
CPPFLAGS=-I$(HOMEBREW_LOC)/opt/gettext/include -I$(LLVM_LOC)/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include

This seemed to solve the issue and allow me to backdate the files as seen above

I was running 1.1.7 before, as I downloaded the package in early November, and it was all working yesterday. I did the require function, and seen it had updated to 1.1.8 and it no longer ran. 


So, I ran the code you gave me and got this error message

Warning messages:
1: In checkMatrixPackageVersion() :
  Package version inconsistency detected.
TMB was built with Matrix version 1.6.5
Current Matrix version is 1.6.1
Please re-install 'TMB' from source using install.packages('TMB', type = 'source') or ask CRAN for a binary version of 'TMB' matching CRAN's 'Matrix' package
2: In checkDepPackageVersion(dep_pkg = "TMB") :
  Package version inconsistency detected.
glmmTMB was built with TMB version 1.9.10
Current TMB version is 1.9.6
Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)

I went ahead and updated both packages; then it seemed to work on a fresh session!

library(glmmTMB)
> set.seed(101); dd <- data.frame(x = 1:10)
> glmmTMB(cbind(x, 10-x) ~ 1, family = betabinomial, data = dd)
Formula:          cbind(x, 10 - x) ~ 1
Data: dd
      AIC       BIC    logLik  df.resid 
 51.17650  51.78167 -23.58825         8 

Number of obs: 10

Dispersion parameter for betabinomial family (): 2.95 

Fixed Effects:

Conditional model:
(Intercept)  
     0.2273  
> sessionInfo()
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.3.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Europe/London
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] glmmTMB_1.1.6

loaded via a namespace (and not attached):
 [1] lme4_1.1-35         codetools_0.2-19    numDeriv_2016.8-1.1 multcomp_1.4-25     Matrix_1.6-5       
 [6] lattice_0.22-5      TH.data_1.1-2       coda_0.19-4.1       splines_4.3.2       estimability_1.4.1 
[11] zoo_1.8-12          emmeans_1.10.0      mvtnorm_1.2-4       TMB_1.9.10          xtable_1.8-4       
[16] nloptr_2.0.3        sandwich_3.1-0      grid_4.3.2          compiler_4.3.2      rstudioapi_0.15.0  
[21] boot_1.3-28.1       tools_4.3.2         nlme_3.1-164        minqa_1.2.6         survival_3.5-7     
[26] Rcpp_1.0.12         MASS_7.3-60.0.1    
> 

When I use my own data again, I get this issue

Error in fitTMB(TMBStruc) : 
  negative log-likelihood is NaN at starting parameter values

how might I go about fixing this? It seems there have been issues with the new version of the matrix; even when I go back to 1.6.2, the same error still shows up. 

thanks!

**Comment by bbolker** [comment]
What happens if you install all the *latest* versions, compiled from compatible versions (Matrix 1.6.5/TMB 1.9.10/glmmTMB 1.1.8)? I'm still not clear on what the original symptoms of "no longer ran" were when you updated to glmmTMB 1.1.8. Did you start getting warnings about incompatible versions, or is this when the "negative log-likelihood is NaN" error started?

If you install everything directly from the main CRAN repo (in a clean R session, no packages loaded)

```r
oo <- options(repos = "https://cran.r-project.org/")
install.packages("Matrix")
install.packages("TMB")
install.packages("glmmTMB")
```

... is it still the case that the minimal example I gave works, but your code doesn't?

It's hard to see anything in the [NEWS file](https://cran.r-project.org/web/packages/glmmTMB/news.html) relating to 1.1.8 that would be a likely culprit ...

**Comment by erinrickman** [comment]
Hello

Yes, they no longer ran, so the original error message included two issues: the beta-binomial could not be found and the incompatible versions.

Error in betabinomial(link = "logit") :
could not find function "betabinomial"
In addition: Warning message:
In checkDepPackageVersion(dep_pkg = "TMB") :
Package version inconsistency detected.
glmmTMB was built with TMB version 1.9.6
Current TMB version is 1.9.10
Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)

I installed the code again as instructed, and when I utilize the library function again, it now gives me the incompatible package massage 

library(glmmTMB)
Warning message:
In checkDepPackageVersion(dep_pkg = "TMB") :
  Package version inconsistency detected.
glmmTMB was built with TMB version 1.9.6
Current TMB version is 1.9.10
Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)

So, I went back and installed the old version of TMB and got this
Error in fitTMB(TMBStruc) : 
  negative log-likelihood is NaN at starting parameter values

proximityface <- glmmTMB::glmmTMB(cbind(convertedtime, convertedtotal) ~ condition*object_relation + (1|subject), 
                                  family = betabinomial(link = "logit"),
                                  data = rawface[rawface$behavior == "proximity",])

My two cbind factors are set as numbers, with no negative values or half numbers as per model requirement. This same code was running yesterday, so no idea!

**Comment by erinrickman** [comment]
So, just because i forgot to pop it in, with the newest TMB function, your code didn't work because it wouldn't load due to incompatible packages, but does run on the older version of TMB where mine throws the NaN error

**Comment by bbolker** [comment]
(I might not have more time until late today or tomorrow.)

Note that the "version inconsistency" is a **warning**, and may be overly cautious. Models will often be fine even when there is a version inconsistency (and, if something fails, it will be obvious - you won't get subtly wrong answers). I am definitely more interested in helping you get the latest version of glmmTMB working than doing archaeology to get some particular past version installed and working (I do understand that your highest priority is getting your code running, but if it's only running with a very particular setup then that's also a fragile solution).

If you can e-mail or post your data I can take a try and make sure that it *can* work on the latest versions (once we get the setup sorted out).

**Comment by erinrickman** [comment]
I'll email my data over to you. Thank you so much. It would be great to get this working on the newer versions.

Yes, I'm on some tight deadlines as I have some papers to submit! So it threw me a little. So, I'm thankful for your speedy response!

On the newer packages, your code is running, but mine still isn't! 

It will run other behaviors in my model, just not the proximity one!

**Comment by erinrickman** [comment]
Here you go!

From: Ben Bolker ***@***.***>
Date: Tuesday, 13 February 2024 at 15:31
To: glmmTMB/glmmTMB ***@***.***>
Cc: erinrickman ***@***.***>, Author ***@***.***>
Subject: Re: [glmmTMB/glmmTMB] glmmTMB cannot find betabinomal function (Issue #988)

(I might not have more time until late today or tomorrow.)

Note that the "version inconsistency" is a warning, and may be overly cautious. Models will often be fine even when there is a version inconsistency (and, if something fails, it will be obvious - you won't get subtly wrong answers). I am definitely more interested in helping you get the latest version of glmmTMB working than doing archaeology to get some particular past version installed and working (I do understand that your highest priority is getting your code running, but if it's only running with a very particular setup then that's also a fragile solution).

If you can e-mail or post your data I can take a try and make sure that it can work on the latest versions (once we get the setup sorted out).

**Comment by bbolker** [comment]
I think you might have failed in your attempt to e-mail me directly with your data/code (you can find my e-mail by googling ...)

**Comment by bbolker** [comment]
You commented offline that the problem was an issue with importing the data. If you can summarize what went wrong (i.e., what kind of incorrect information you tried to pass to `glmmTMB` maybe we can add a check/warning that will make the problem more obvious (clearly we can't check for *everything* that could have gone wrong in the data, but it would be good to have a more informative error message than "NaN at starting values")

**Comment by erinrickman** [comment]
Hello!

So, I have thousands of rows of data, and it turned out that there was 1 negative value that I accidentally put in, which wasn't an easy find! So, maybe a warning if negative values are detected, as I know the NaN issue has also been related to matrix issues before(which was the rabbit hole I went down!) It seems that because this was missed and I trolled through forums, I went the wrong way and created some of my own issues, haha! So maybe more specific around a negative value dected and non-intergers if there isn't already!

While the Makevars were needed for devtools, it may(or may not) have made a difference for this. Maybe there were package clashes before this, or I clicked something I shouldn't have done (though it all seems in order and working now with all new packages!)

Thanks for your time!

**Comment by bbolker** [comment]
I thought we were detecting this properly, but apparently not: `family = betabinomial` gives an inscrutable error, `family = binomial` returns crazy answers, but `glm(..., family = binomial)` gives an informative error.  So this is definitely a hole to patch.

```r
library(glmmTMB)
set.seed(101); dd <- data.frame(x = 1:10)
m1 <- glmmTMB(cbind(x, 10-x) ~ 1, family = betabinomial, data = dd)
sessionInfo()

dd2 <- rbind(dd, data.frame(x=-1))
update(m1, data = dd2)  ## NOT detected
update(m1, data = dd2, family = binomial)  ## gives weird/infinite answers

glm(cbind(x, 10-x) ~ 1, family = binomial, data = dd2)
##Error in family$linkfun(mustart) : Value -0.0454545 out of range (0, 1)
```

**Comment by erinrickman** [comment]
So, to be really silly and dumb it down for someone who don't know stats to well, what does this mean for the coding and the outputs?

**Comment by bbolker** [comment]
All I'm saying here is that we should be doing a better job detecting incorrect input. Negative counts is the main thing that can go wrong, and undetected, if you specify the data in the 2-column `cbind(# successes, #failures)` format. It would be nice to warn people if they were accidentally specifying `cbind(# successes, total)` instead &mdash; a common mistake &mdash; but there's no way to detect that for sure (you can rule it out if the first column is ever greater than the second column, but it wouldn't be unusual for the observed proportions to always be < 0.5 [i.e., first column always *less than* the second column], so there's no way to know ...

**Comment by erinrickman** [comment]
Ah! That makes sense!

**Comment by bbolker** [comment]
There's a proof of concept [here](https://github.com/glmmTMB/glmmTMB/blob/check_neg_binom_vals/misc/neg_binom_ex.R) for how we can extend the existing `$initialize` expression for binomial models (we can adjust the `betabinomial` family ourself, we'll need a little bit of special-case code in the body of `glmmTMB()` to hack `binomial` families appropriately)

**Comment by bbolker** [comment]
Resolved by merging #1002
